Modules/_pickle.c

```diff
@@ -1678,25 +1678,40 @@ _Unpickler_SetInputStream(UnpicklerObject *self, PyObject *file)
 {
     /* Optional file methods */
     if (_PyObject_LookupAttr(file, &_Py_ID(peek), &self->peek) < 0) {
-        return -1;
+        goto error;
     }
     if (_PyObject_LookupAttr(file, &_Py_ID(readinto), &self->readinto) < 0) {
-        return -1;
+        goto error;
     }
-    (void)_PyObject_LookupAttr(file, &_Py_ID(read), &self->read);
-    (void)_PyObject_LookupAttr(file, &_Py_ID(readline), &self->readline);
-    if (!self->readline || !self->read) {
-        if (!PyErr_Occurred()) {
-            PyErr_SetString(PyExc_TypeError,
-                            "file must have 'read' and 'readline' attributes");
+    int no_read_attr = 0;
+    if (_PyObject_LookupAttr(file, &_Py_ID(read), &self->read) < 0) {
+        if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
```

Comment:
It never matches. `_PyObject_LookupAttr()` converts `AttributeError` into returning value 0.

Reply:
AFAICS, 2d6a7b5e693e4ba0cbe42e0847510454318c2062 should be correct:

- if -1 is returned, clear the exception
- use the `result` output variable instead of temporary `no_read*_attr` variables